### DISCUSSION TITLE
Segfault on macOS after manual packing with PyArmor

### DISCUSSION THREAD

**Discussion by Adir-Shemesh** [dashingsoft/pyarmor#1229] [Q&A]
We followed the steps detailed in [this issue](https://github.com/dashingsoft/pyarmor/discussions/1107#discussioncomment-5557571) (specifically because of the error message that appears in the linked comment), and [the official documentation](https://pyarmor.readthedocs.io/en/stable/topic/repack.html#packing-obfuscated-scripts-manually).

The result works fine on Windows and Linux, but on macOS it causes a segfault.

**Comment by jondy** [comment]
How about a hello world script?
And please report Pyarmor version, PyInstaller version, and MacOS information (Intel or M1).

**Comment by Adir-Shemesh** [comment]
This reproduces with a hello world script.
pyarmor==8.1.9 
pyinstaller==5.8.0
not sure about the processor, using the github-hosted macos runner

**Comment by Adir-Shemesh** [comment]
@jondy are there any further steps I can take? 🙂

**Comment by jondy** [comment]
@Adir-Shemesh it will be fixed in the next release 8.2.1, I have found the problem.

Also refer to #1243, I think the reason is same for #1243 and this

**Comment by jondy** [comment]
Fixed in 8.2.1, please upgrading to the latest version.

**Reply by Adir-Shemesh** [reply]
packing manually still segfaults on macos, using `--pack` still results in `[5303] Error loading Python lib '/tmp/_MEIHBP838/libpython3.11.so.1.0': dlopen: /tmp/_MEIHBP838/libpython3.11.so.1.0: cannot open shared object file: No such 
file or directory` on linux :(

**Comment by Alias419** [comment]
I am having similar issue, I am able to build on windows 64bit, MacOS M1, however MacOS 64 bit fails with segment fault. On Mac OS 64 bit I am able to obfuscate and run, or just pack with pyinstaller and run. Just not able to pack obfuscate or obfuscate then pack. Both ways give me segment fault. Pyarmor 8.2.1 Pro, Pyinstaller 5.11.0, Also reproduced with Hello World

**Comment by jondy** [comment]
@Alias419 how about apply this patch?
ff7e0e3e3c9b2ad1ab28ba023e5d7db4c26a9067

**Reply by Alias419** [reply]
Still getting segmentation fault. However that was obfuscate then pack. Repacked worked or so I thought. Spec file used to pack creates two exe's. The one specified with --pack still has seg fault. The one outside dist/obf runs but not obfuscated.

**Comment by jondy** [comment]
Does it still segment fault?
```
pyinstaller foo.py
pyarmor gen --pack dist/foo/foo foo.py
dist/foo/foo
```

Make sure there is no warning logs when run the second command `pyarmor gen`

**Reply by Alias419** [reply]
Removed all options, and used just gen --pack. With that throws error: 'Security Auth Failure: make sure python is signed with codesign util' for keyring backend. Reinstalled keyring, back to seg fault error. Build runs fine on Mac OS M1 chipset and Windows x64

**Reply by jondy** [reply]
Well, check code sign of `dist/foo/foo` and `find dist/foo -name pyarmor_runtime.so`
```
codesign -v dist/foo/foo
codesign -v $(find dist/foo/ -name pyarmor_runtime.so)
```
And codesign them if it's invalid

**Reply by Alias419** [reply]
The keyring backend pointed me to recreating my coding environment with pyenv vs pipenv. I am now able to build using the following args:
pyarmor gen -O dist/obf -r --obf-code 1 --enable-jit --mix-str ui.py

After which i can pack with pyinstaller spec and get a functional exe. However if I swap to the following args:
pyarmor gen -O dist/obf -r --obf-code 1 --mix-str --enable-rft --enable-bcc ui.py

I now get a return _run_code(code, main_globals, None,

**Reply by Alias419** [reply]
> Does it still segment fault?
> 
> ```
> pyinstaller foo.py
> pyarmor gen --pack dist/foo/foo foo.py
> dist/foo/foo
> ```
> 
> Make sure there is no warning logs when run the second command `pyarmor gen`

Still working to resolve, today after starting a fresh virtual env, after pyinstaller --onefile, app works fine. I have gotten new error when trying to pack with pyarmor. Error below:
install_name_tool -change @rpath/lib/libpython3.11.dylib @rpath/Python .pyarmor/pack/dist/pyarmor_runtime_005133/pyarmor_runtime.so
INFO     re-signing "pyarmor_runtime.so"
ERROR    sign_binary() got an unexpected keyword argument 'identify'

**Reply by jondy** [reply]
@Alias419 apply to this patch 1fe690cc581a5d887d1a99dbc7eec9aade30c07f to fix this error

**Reply by Alias419** [reply]
Applied patch and had to resign, giving error below:

(-25293, 'Security Auth Failure: make sure python is signed with codesign util')

**Reply by jondy** [reply]
Please check the log and refer to https://pyarmor.readthedocs.io/en/stable/questions.html#segment-fault-in-apple-m1 to fix it
